### PR TITLE
[UDX-1216] create custom partitioner, 'UdxStreamPartitioner', for msk connect connector

### PR DESCRIPTION
**Pull request by inbrewj-udc**
## Description

- Adds a new package, `io.confluent.connect.s3.extensions` and a new class, `UdxStreamPartitioner` that creates a partition structure in S3 derived from: 
    - the streamUuid (in the header of each kafka message)
    - fields from the JSON payload
- Also includes a suite of unit tests that _should_ ensure that even if the message and payload are completely corrupted, the message will still land in a predictable place in our S3 data lake 
    - For the exact behaviour, see `UdxStreamPartitionerTest.java`
- I'm hardly an experienced Java programmer but I hope what I've written is readable 🤞 
- Relies heavily on `com.fasterxml.jackson` for JSON processing and `org.joda.time` for timestamp processing. These libs were already used heavily elsewhere in the connector; I didn't want to introduce anything un-idiotmatic
- Adds a Github workflow to test that packaging up the connector works and the unit tests I've written run properly (and pass)
    - Caveat: this workflow currently takes 20 minutes 😱 
    - Unfortunately, the bulk of the time spent in this workflow is sunk into downloading / building / installing a few dependencies from source. Not _all_ dependencies, mind you, just five or so that are also authored by Confluent Inc. One of these happens to be Kafka itself; another 😱. These are described in the script: `install_upstream_dependencies.sh`
    - There might be some way to optimise this process by hosting the prebuilt jars ourselves, doing something clever with GitHub actions to cache the dependencies or doing something clever with a docker container we could host ourselves?
    - Truth be told there might also be some flag to add to maven to speed things up somehow - this entire toolchain is very new to me
- Also adds a Github action workflow to push the ZIPped up build to an S3 bucket visible by MSK Connect. Adds a section in the README that explains exactly how to update the the msk connect setup     

## How have I tested this?

- After building a `.zip` archive and pushing this connector to a valid msk_connect_connector config on our `dev` account(see [this PR in udx-infra](https://github.com/Urban-Data-Collective/udx-infra/pull/155)), I essentially tailed the cloudwatch logs with this command and ensured there were no logs describing horrible Java exceptions:
    - `aws logs tail /msk-connect/udx-dev-s3-sink --follow --region eu-west-1`

An example output snippet of the logs:

```
2022-06-23T15:01:53+00:00 udx-dev-mskconnect-to-s3-data-lake-617e4369-2901-41a8-a326-c9aa31842666-2 [Worker-0130ecaa298b386bd] [2022-06-23 15:01:53,082] INFO [udx-dev-mskconnect-to-s3-data-lake|task-0] Opening record writer for: topics/aws.db.data-streams.records.0/streamUuid=a36d91b1-0365-4e4d-b798-f6d1b3ddf112/entityId=myway-street-076/2022-06/day=23/hour=15/aws.db.data-streams.records.0+9+0000714781.avro (io.confluent.connect.s3.format.avro.AvroRecordWriterProvider:74)
2022-06-23T15:01:54+00:00 udx-dev-mskconnect-to-s3-data-lake-617e4369-2901-41a8-a326-c9aa31842666-2 [Worker-0130ecaa298b386bd] [2022-06-23 15:01:54,273] INFO [udx-dev-mskconnect-to-s3-data-lake|task-0] encoding partition with UdxStreamPartitioner... (io.confluent.connect.s3.extensions.UdxStreamPartitioner:136)
2022-06-23T15:01:54+00:00 udx-dev-mskconnect-to-s3-data-lake-617e4369-2901-41a8-a326-c9aa31842666-2 [Worker-0130ecaa298b386bd] [2022-06-23 15:01:54,273] INFO [udx-dev-mskconnect-to-s3-data-lake|task-0] Parsing value... (io.confluent.connect.s3.extensions.UdxStreamPartitioner:137)
2022-06-23T15:01:54+00:00 udx-dev-mskconnect-to-s3-data-lake-617e4369-2901-41a8-a326-c9aa31842666-2 [Worker-0130ecaa298b386bd] [2022-06-23 15:01:54,273] INFO [udx-dev-mskconnect-to-s3-data-lake|task-0] Value: {"id":"myway-street-063","received_at":"2022-06-23T15:01:54.022933023Z","timestamp":"2022-06-23T15:01:54.268Z","data":{"humidity":57,"light":2309,"motion":18,"temperature":26.8,"vdd":3663}} (io.confluent.connect.s3.extensions.UdxStreamPartitioner:140)
2022-06-23T15:01:54+00:00 udx-dev-mskconnect-to-s3-data-lake-617e4369-2901-41a8-a326-c9aa31842666-2 [Worker-0130ecaa298b386bd] [2022-06-23 15:01:54,273] INFO [udx-dev-mskconnect-to-s3-data-lake|task-0] Assuming streamUuid is in the headers... (io.confluent.connect.s3.extensions.UdxStreamPartitioner:144)
2022-06-23T15:01:54+00:00 udx-dev-mskconnect-to-s3-data-lake-617e4369-2901-41a8-a326-c9aa31842666-2 [Worker-0130ecaa298b386bd] [2022-06-23 15:01:54,273] INFO [udx-dev-mskconnect-to-s3-data-lake|task-0] Getting offering_uuid value from headers... (io.confluent.connect.s3.extensions.UdxStreamPartitioner:52)
2022-06-23T15:01:54+00:00 udx-dev-mskconnect-to-s3-data-lake-617e4369-2901-41a8-a326-c9aa31842666-2 [Worker-0130ecaa298b386bd] header key => offering_uuidheader value => a36d91b1-0365-4e4d-b798-f6d1b3ddf112
2022-06-23T15:01:54+00:00 udx-dev-mskconnect-to-s3-data-lake-617e4369-2901-41a8-a326-c9aa31842666-2 [Worker-0130ecaa298b386bd] [2022-06-23 15:01:54,273] INFO [udx-dev-mskconnect-to-s3-data-lake|task-0] setting streamUuid to: a36d91b1-0365-4e4d-b798-f6d1b3ddf112 (io.confluent.connect.s3.extensions.UdxStreamPartitioner:61)
2022-06-23T15:01:54+00:00 udx-dev-mskconnect-to-s3-data-lake-617e4369-2901-41a8-a326-c9aa31842666-2 [Worker-0130ecaa298b386bd] [2022-06-23 15:01:54,274] WARN [udx-dev-mskconnect-to-s3-data-lake|task-0] Could not parse payload into location POJO (io.confluent.connect.s3.extensions.UdxStreamPartitioner:184)
2022-06-23T15:01:54+00:00 udx-dev-mskconnect-to-s3-data-lake-617e4369-2901-41a8-a326-c9aa31842666-2 [Worker-0130ecaa298b386bd] [2022-06-23 15:01:54,274] INFO [udx-dev-mskconnect-to-s3-data-lake|task-0] Mapped to: class io.confluent.connect.s3.extensions.FlatTimestampPayload (io.confluent.connect.s3.extensions.UdxStreamPartitioner:187)
2022-06-23T15:01:54+00:00 udx-dev-mskconnect-to-s3-data-lake-617e4369-2901-41a8-a326-c9aa31842666-2 [Worker-0130ecaa298b386bd] [2022-06-23 15:01:54,274] INFO [udx-dev-mskconnect-to-s3-data-lake|task-0] Mapping record value into object... (io.confluent.connect.s3.extensions.UdxStreamPartitioner:200)
2022-06-23T15:01:54+00:00 udx-dev-mskconnect-to-s3-data-lake-617e4369-2901-41a8-a326-c9aa31842666-2 [Worker-0130ecaa298b386bd] [2022-06-23 15:01:54,274] INFO [udx-dev-mskconnect-to-s3-data-lake|task-0] OcpiPayload [ entityId: myway-street-063, timestamp: 2022-06-23T15:01:54.268Z ] (io.confluent.connect.s3.extensions.UdxStreamPartitioner:201)
2022-06-23T15:01:54+00:00 udx-dev-mskconnect-to-s3-data-lake-617e4369-2901-41a8-a326-c9aa31842666-2 [Worker-0130ecaa298b386bd] [2022-06-23 15:01:54,274] INFO [udx-dev-mskconnect-to-s3-data-lake|task-0] parsing timestamp : 2022-06-23T15:01:54.268Z (io.confluent.connect.s3.extensions.UdxStreamPartitioner:78)
2022-06-23T15:01:54+00:00 udx-dev-mskconnect-to-s3-data-lake-617e4369-2901-41a8-a326-c9aa31842666-2 [Worker-0130ecaa298b386bd] [2022-06-23 15:01:54,274] INFO [udx-dev-mskconnect-to-s3-data-lake|task-0] Opening record writer for: topics/aws.db.data-streams.records.0/streamUuid=a36d91b1-0365-4e4d-b798-f6d1b3ddf112/entityId=myway-street-063/2022-06/day=23/hour=15/aws.db.data-streams.records.0+44+0000714738.avro (io.confluent.connect.s3.format.avro.AvroRecordWriterProvider:74)
```

- If there were consistent exceptions the connector would fail to start. However, this iteration of the partitioner seems to run ok as evidenced by an MSK Connect Connector in a 'running' state:

image

- ...and also evidenced by the desired file / partition structure forming nicely in S3:

image

image

Looking at this particular file with the avro2json script in the infra repo yields:

```
avro2json_from_s3_file_path "s3://udx-dev-streams-data-lake/topics/aws.db.data-streams.records.0/streamUuid=9a2531d2-9999-4fab-b143-f04083b9871e/entityId=GB*CKL*E1507/2022-06/day=23/hour=14/aws.db.data-streams.records.0+18+0001128387.avro"
Getting avro file from s3://udx-dev-streams-data-lake/topics/aws.db.data-streams.records.0/streamUuid=9a2531d2-9999-4fab-b143-f04083b9871e/entityId=GB*CKL*E1507/2022-06/day=23/hour=14/aws.db.data-streams.records.0+18+0001128387.avro
download: s3://udx-dev-streams-data-lake/topics/aws.db.data-streams.records.0/streamUuid=9a2531d2-9999-4fab-b143-f04083b9871e/entityId=GB*CKL*E1507/2022-06/day=23/hour=14/aws.db.data-streams.records.0+18+0001128387.avro to ../../../../../../../../tmp/fromS3.avro
22/06/23 16:06:52 WARN util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
{
  "string": "{\"id\":\"GB*CKL*E1507\",\"evseId\":\"GB*CKL*E1507\",\"timestamp\":\"2022-06-23T14:45:53Z\",\"type\":\"EVChargingSession\",\"provider\":\"CKL\",\"partyId\":\"CKL\",\"countryCode\":\"GB\",\"sessionId\":48181,\"connectorId\":1111,\"address\":{\"addressCountry\":\"GB\",\"addressLocality\":\"Martlesham\",\"postalCode\":\"IP5 3RE\",\"streetAddress\":\"Martlesham Heath, Martlesham, Ipswich IP5 3RE, UK\"},\"location\":{\"coordinates\":[1.2855696,52.0599247],\"type\":\"Point\"},\"sessionStartTime\":\"2022-06-23T08:06:12Z\",\"sessionEndTime\":\"2022-06-23T14:45:47Z\",\"totalCost\":{\"exclVat\":14.39,\"inclVat\":17.27},\"sessionDurationMins\":399.6,\"chargingDurationMins\":399.6,\"totalKW\":49.354}"
}
```

##### Testing done:
- [x] Unit tests
- [ ] Integration tests
- [ ] System tests
- [x] Manual tests

### CODE REVIEW

File: .github/workflows/package-and-test.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        uses: actions/setup-java@v3
+        with:
+          java-version: '11'
+          distribution: 'zulu'
+      - name: Install upstream dependencies
+        run: ./install_upstream_dependencies.sh
+      - name: Package up ZIP
+        run: mvn --batch-mode package -DskipTests
+      - name: Run UdxStreamPartitioner unit tests
+        run: mvn --batch-mode -Dtest=UdxStreamPartitionerTest -Dsurefire.failIfNoSpecifiedTests=false clean test
```
Comment (sarahgudc) [inline .github/workflows/package-and-test.yml]: I think we could also add a deploy step to upload the latest built .zip to the S3 bucket (a similar pattern to what we already do with docker images and uploading new ones to ECR

File: build_zip_and_copy_to_infra.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+# e.g. ~/workshop/udx-infra/terraform/modules/data-lake
+PATH_TO_INFRA=$1
+OUTPUT_ZIP_NAME="confluentinc-kafka-connect-s3-10.0.8-UdxStreamPartitioner.zip"
+
+mvn clean install -DskipTests && \
+echo "Removing existing build in infra, just to be safe..." && \
+rm -f "$PATH_TO_INFRA/$OUTPUT_ZIP_NAME" || true && \
+echo "Copying ZIP archived build to $PATH_TO_INFRA" && \
+cp kafka-connect-s3/target/components/packages/confluentinc-kafka-connect-s3-10.0.8.zip "$PATH_TO_INFRA/$OUTPUT_ZIP_NAME"
```
Comment (sarahgudc) [inline build_zip_and_copy_to_infra.sh]: perhaps don't need to copy to infra
Instead we could 'deploy' this to the S3 bucket created in terraform

Comment (inbrewj-udc) [reply build_zip_and_copy_to_infra.sh]: 🙇 🙇 I've added a run down of what I need to do in the PR description 👌

File: .github/workflows/package-and-test.yml (hunk 1 above)
Comment (inbrewj-udc) [reply .github/workflows/package-and-test.yml]: All done ✅

File: build_zip_and_copy_to_infra.sh (hunk 2 above)
Comment (inbrewj-udc) [reply build_zip_and_copy_to_infra.sh]: All done ✅

File: bin/build_zip_and_copy_to_infra.sh [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#! /bin/bash
+
+# e.g. ~/workshop/udx-infra/terraform/modules/data-lake
```
Comment (sarahgudc) [inline bin/build_zip_and_copy_to_infra.sh]: Do we need this script anymore if we aren't storing the build in the infra repo?

File: bin/copy_ci_build_to_folder.sh [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#! /bin/bash
+
```
Comment (sarahgudc) [inline bin/copy_ci_build_to_folder.sh:2]: same for this script - do we still need this?

**Review by sarahgudc — APPROVED** [review]
LGTM! - just 2 questions about scripts

File: bin/copy_ci_build_to_folder.sh (hunk 4 above)
Comment (inbrewj-udc) [reply bin/copy_ci_build_to_folder.sh:2]: Yep, we still need this one - this particular script renames the ZIP artifact created on CI to `confluentinc-kafka-connect-s3-UdxStreamPartitioner.zip` which means we can reference a constant filename in [the terraform config](https://github.com/Urban-Data-Collective/udx-infra/blob/UDX-1216-create-custom-partitioner-for-msk-connect-connector/terraform/modules/data-lake/variables.tf#L31). This one is called in the Github workflows

File: bin/build_zip_and_copy_to_infra.sh (hunk 3 above)
Comment (inbrewj-udc) [reply bin/build_zip_and_copy_to_infra.sh]: Yep, this is entirely superflous

**Pull request merged**